component/src/main/java/org/wso2/extension/siddhi/execution/streamingml/clustering/kmeans/KMeansIncrementalSPExtension.java

```diff
@@ -0,0 +1,305 @@
+                    throw new SiddhiAppValidationException("numberOfClusters has to be a constant.");
+                }
+                if (attributeExpressionExecutors[2].getReturnType() == Attribute.Type.INT) {
+                    coordinateStartIndex = 3;
+                    numberOfClusters = (Integer) ((ConstantExpressionExecutor)
+                            attributeExpressionExecutors[2]).getValue();
+                } else {
+                    throw new SiddhiAppValidationException("numberOfClusters should be of type int but found " +
+                            attributeExpressionExecutors[2].getReturnType());
+                }
+            } else {
+                throw new SiddhiAppValidationException("decayRate has to be a constant.");
+            }
+        } else {
+            throw new SiddhiAppValidationException("The second query parameter should either be decayRate or " +
+                    "numberOfClusters which should be of type float or int respectively but found " +
+                    attributeExpressionExecutors[1].getReturnType());
+        }
+
+        dimensionality = attributeExpressionExecutors.length - coordinateStartIndex;
+        coordinateValuesOfCurrentDataPoint = new double[dimensionality];
+
+        //validating all the attributes to be variables
+        for (int i = coordinateStartIndex; i < coordinateStartIndex + dimensionality; i++) {
+            if (!(this.attributeExpressionExecutors[i] instanceof VariableExpressionExecutor)) {
+                throw new SiddhiAppValidationException("The attributes should be variable but found a " +
+                        this.attributeExpressionExecutors[i].getClass().getCanonicalName());
+            }
+        }
+
+        String siddhiAppName = siddhiAppContext.getName();
+        modelName = modelName + "." + siddhiAppName;
+        if (logger.isDebugEnabled()) {
+            logger.debug("model name is " + modelName);
+        }
+        int maxIterations = 2;
+        clusterer = new Clusterer(numberOfClusters, maxIterations, modelName, siddhiAppName, dimensionality);
+
+        List<Attribute> attributeList = new ArrayList<>(1 + dimensionality);
+        attributeList.add(new Attribute("euclideanDistanceToClosestCentroid", Attribute.Type.DOUBLE));
+        for (int i = 1; i <= dimensionality; i++) {
+            attributeList.add(new Attribute("closestCentroidCoordinate" + i, Attribute.Type.DOUBLE));
+        }
+        return attributeList;
+    }
+
+    @Override
+    protected void process(ComplexEventChunk<StreamEvent> streamEventChunk, Processor processor,
+                           StreamEventCloner streamEventCloner, ComplexEventPopulater complexEventPopulater) {
+        synchronized (this) {
+            while (streamEventChunk.hasNext()) {
+                StreamEvent streamEvent = streamEventChunk.next();
+
+                numberOfEventsReceived++;
```

Comment:
Why are we counting the events here?

Reply:
just to make sure all the events are captured and processed when testing with large datasets. no need in the production code

Reply:
i will add the increment also inside if(isdebugenabled). is that okay?